File: 2022/day10/ruby/solution.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'set'

def simulate_cpu(instructions)
  return enum_for(:simulate_cpu, instructions) unless block_given?

  x = 1
  cycle = 0

  instructions.each do |line|
    if line == 'noop'
      cycle += 1
      yield cycle, x
    else
      v = line.split[1].to_i
      cycle += 1
      yield cycle, x
      cycle += 1
      yield cycle, x
      x += v
    end
  end
end

def part1(instructions)
  target_cycles = [20, 60, 100, 140, 180, 220].to_set
  total = 0

  simulate_cpu(instructions) do |cycle, x|
    total += cycle * x if target_cycles.include?(cycle)
  end

  total
end

def part2(instructions)
  screen = []
  row = []

  simulate_cpu(instructions) do |cycle, x|
    pos = (cycle - 1) % 40
    row << ((pos - x).abs <= 1 ? '#' : '.')

    if cycle % 40 == 0
      screen << row.join
      row = []
    end
  end

  screen.join("\n")
end

def main
  script_dir = File.dirname(File.expand_path(__FILE__))
  input_file = File.join(script_dir, '..', 'input.txt')

  instructions = File.read(input_file).strip.split("\n")

  puts "Part 1: #{part1(instructions)}"
  puts 'Part 2:'
  puts part2(instructions)
end

main if __FILE__ == $PROGRAM_NAME

File: 2022/day10/node/solution.js
import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function* simulateCpu(instructions) {
  let x = 1;
  let cycle = 0;

  for (const line of instructions) {
    if (line === 'noop') {
      cycle++;
      yield [cycle, x];
    } else {
      const v = parseInt(line.split(' ')[1], 10);
      cycle++;
      yield [cycle, x];
      cycle++;
      yield [cycle, x];
      x += v;
    }
  }
}

function part1(instructions) {
  const targetCycles = new Set([20, 60, 100, 140, 180, 220]);
  let total = 0;

  for (const [cycle, x] of simulateCpu(instructions)) {
    if (targetCycles.has(cycle)) {
      total += cycle * x;
    }
  }

  return total;
}

function part2(instructions) {
  const screen = [];
  let row = [];

  for (const [cycle, x] of simulateCpu(instructions)) {
    const pos = (cycle - 1) % 40;
    if (Math.abs(pos - x) <= 1) {
      row.push('#');
    } else {
      row.push('.');
    }

    if (cycle % 40 === 0) {
      screen.push(row.join(''));
      row = [];
    }
  }

  return screen.join('\n');
}

const inputFile = join(__dirname, '..', 'input.txt');
const instructions = readFileSync(inputFile, 'utf-8').trim().split('\n');

console.log('Part 1:', part1(instructions));
console.log('Part 2:');
console.log(part2(instructions));
